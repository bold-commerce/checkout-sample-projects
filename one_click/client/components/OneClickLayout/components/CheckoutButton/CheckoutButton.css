.CheckoutButton__Container {
  border-top: 1px solid;
  border-color: rgba(0, 0, 0, 0.4);
}
.CheckoutButton {
  color: white;
  width: 100%;
  margin: 16px 0;
  background-color: #0a75c2;
}

.CheckoutButton:disabled {
  background-color: rgba(0, 0, 0, 0.65);
}
