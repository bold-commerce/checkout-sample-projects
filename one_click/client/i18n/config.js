import i18n from 'i18next';
import enTranslation from './en/translation.json';
import { initReactI18next } from 'react-i18next';

i18n
  .use(initReactI18next)
  .init({
    resources: {
      en: {
        translation: enTranslation
      }
    },
    lng: 'en',
    fallbackLng: 'en',
    interpolation: {
      escapeValue: false, 
    }
  });

export default i18n;