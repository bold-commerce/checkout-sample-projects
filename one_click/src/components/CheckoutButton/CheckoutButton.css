button.CheckoutButton,
button.CheckoutButton:hover {
  color: white;
  width: 100%;
  margin: 16px 0;
  background-color: #0a75c2;    
  padding: 8px;
  border-radius: 4px;
  box-shadow: 0px 1px 2px 0px #0000001a;
  border: 1px solid #0000001A;
}

button.CheckoutButton:disabled {
  background-color: rgba(0, 0, 0, 0.65);
}

button.CheckoutButton__Show {
  visibility: visible;
}

button.CheckoutButton__Hide {
  display:none;
  visibility: hidden;
}

.CheckoutButton__Desktop__Wrapper {
  padding: 0px 32px;
}

@media only screen and (max-width: 769px) {
  button.CheckoutButton__Desktop {
    display: none;
  }
}

@media only screen and (min-width: 770px) {
  button.CheckoutButton__Mobile {
    display: none;
  }
}
