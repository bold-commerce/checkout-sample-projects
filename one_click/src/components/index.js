export * from './Address';
export * from './BackButton';
export * from './Billing';
export * from './BillingAddress';
export * from './Card';
export * from './CheckoutButton';
export * from './CheckoutForm';
export * from './Confirmation';
export * from './CustomerInfo';
export * from './Discount';
export * from './EmptyState';
export * from './Header';
export * from './Icons';
export * from './Index';
export * from './InventoryItem';
export * from './LineItems';
export * from './LoadingState';
export * from './Payment';
export * from '../pages/Processing';
export * from './Product';
export * from './RadioItem';
export * from './Shipping';
export * from './ShippingLines';
export * from './Summary';
