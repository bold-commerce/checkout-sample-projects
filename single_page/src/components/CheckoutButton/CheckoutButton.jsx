import { useBillingAddress, useCheckoutStore, useLineItems, usePaymentIframe } from '@boldcommerce/checkout-react-components';
import {
  useNavigate,
} from "react-router-dom";
import { Button } from '@boldcommerce/stacks-ui';
import React, { memo, useCallback, useState } from 'react';
import { useAnalytics, useErrorLogging } from '../../hooks';
import { useTranslation } from 'react-i18next';

const CheckoutButton = () => {
  const { processPaymentIframe } = usePaymentIframe();
  const { state } = useCheckoutStore();
  const { data: lineItems } = useLineItems();
  const { data: billingAddress } = useBillingAddress();
  const { orderStatus } = state.orderInfo;
  const processing = orderStatus === 'processing' || orderStatus === 'authorizing';

  return (
    <MemoizedCheckoutButton
      processOrder={processPaymentIframe}
      processing={processing}
      hasBillingAddress={!!billingAddress?.country_code}
      lineItems={lineItems}
      appLoading={state.loadingStatus.isLoading}
    />
  );
};

const MemoizedCheckoutButton = memo(({
  processOrder,
  processing,
  hasBillingAddress,
  lineItems,
  appLoading,
}) => {
  const trackEvent = useAnalytics();
  const logError = useErrorLogging();
  const navigate = useNavigate();
  const [loading, setLoading] = useState(false);
  const { t } = useTranslation();

  const validateInventory = useCallback(async () => {
    const variants = lineItems.map((lineItem) => lineItem.product_data.variant_id).join(',');
    const response = await fetch(`${process.env.INVENTORY_URL}?variants=${variants}`);
    const responseData = await response.json();
    let inventory = Array.from(responseData).reduce((acc,curr)=> (acc[curr.platform_id]=curr.inventory_quantity,acc),{});
    let inventoryIssues = false;

    lineItems.forEach((lineItem) => {
      inventory[lineItem.product_data.variant_id] -= lineItem.product_data.quantity;
      if (inventory[lineItem.product_data.variant_id] < 0) {
        inventoryIssues = true;
      }
    });

    if (inventoryIssues) {
      return responseData.inventory;
    } else {
      return null;
    }
  }, [lineItems]);

  const handleProcessPayment = useCallback(async () => {
    trackEvent('click_complete_order');
    setLoading(true);
    try {
      const inventoryIssues = await validateInventory();
      if (inventoryIssues) {
        setLoading(false);
        navigate('/inventory_issues', { state: inventoryIssues });
      } else {
        setLoading(false);
        await processOrder();
      }
    } catch(e) {
      logError('process_order', e);
      setLoading(false);
    }
  }, [logError, trackEvent, processOrder, validateInventory]);

  return (
    <Button
      type="button"
      className="Checkout__ConfirmButton"
      onClick={handleProcessPayment}
      loading={loading || processing}
      disabled={!hasBillingAddress || processing || appLoading}
    >
      {t('complete_order')}
    </Button>
  );
});

export default CheckoutButton;
