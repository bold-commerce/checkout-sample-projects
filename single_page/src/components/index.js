export * from './Address';
export * from './BillingAddress';
export * from './CheckoutButton';
export * from './CheckoutSection';
export * from './Customer';
export * from './DiscountForm';
export * from './EmptyState';
export * from './Header';
export * from './LineItems';
export * from './LoadingState';
export * from './OrderErrors';
export * from './OrderSummary';
export * from './PaymentMethod';
export * from './ShippingAddress';
export * from './ShippingLines';
