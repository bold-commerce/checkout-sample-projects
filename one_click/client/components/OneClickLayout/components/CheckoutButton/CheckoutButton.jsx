import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { useLocation, useHistory } from 'react-router';
import {
  Button,
  Message,
} from '@boldcommerce/stacks-ui';
import { useCheckoutStore, usePaymentIframe, useLineItems } from '@boldcommerce/checkout-react-components';
import { useInventory } from '../../../../hooks';
import './CheckoutButton.css';

const CheckoutButton = ({ disabled, onClick, loading, className, errorMessage }) => (
  <>
    { errorMessage ? <Message type="alert">{ errorMessage }</Message> : null }
    <div className="CheckoutButton__Container">
      <Button onClick={onClick} disabled={disabled} loading={loading} className={className}>
        Complete Order
      </Button>
    </div>
  </>
);

CheckoutButton.propTypes = {
  disabled: PropTypes.bool,
  onClick: PropTypes.func,
  loading: PropTypes.bool,
  className: PropTypes.string,
  errorMessage: PropTypes.string,
};

const CheckoutButtonContainer = ({ className }) => {
  const { state } = useCheckoutStore();
  const [loading, setLoading] = useState();
  const location = useLocation();
  const { orderStatus } = state.orderInfo;
  const { processPaymentIframe } = usePaymentIframe();
  const checkInventory = useInventory();
  const { data: lineItems } = useLineItems();
  const history = useHistory();

  const orderErrorMessage = state.errors.order?.public_order_id;

  // Don't show the checkout button on the inventory issues page
  const inventoryLocation = location.pathname === '/inventory';

  // don't disable checkout button if only error is order error
  const hasErrors = Object.keys(state.errors).some((errorKey) => errorKey != 'order' && state.errors[errorKey] != null);
  const missingAddress = Object.values(state.applicationState.addresses).some(x => x === null || x.length === 0);
  const checkoutButtonDisabled = state.loadingStatus.isLoading || hasErrors ||  missingAddress;
  const processing = orderStatus === 'processing' || orderStatus === 'authorizing';
  const handleCheckout = async() => {
    setLoading(true);
    try {
      const inventoryIssues = await checkInventory(lineItems);
      if (!inventoryIssues) {
        await processPaymentIframe();
        setLoading(false);
      } else {
        setLoading(false);
        history.push('/inventory', inventoryIssues);
      }
    } catch(e) {  
      setLoading(false);
    }
  };
  const onClick = processing ? null : handleCheckout;

  return inventoryLocation ? null : <CheckoutButton disabled={checkoutButtonDisabled} onClick={onClick} loading={processing || loading} className={className} errorMessage={orderErrorMessage} />;
};

export default CheckoutButtonContainer;
