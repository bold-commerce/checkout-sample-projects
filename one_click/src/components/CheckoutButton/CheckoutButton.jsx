import React, { useState } from 'react';
import PropTypes from 'prop-types';
import {  useHistory } from 'react-router';
import { Button, Message } from '@boldcommerce/stacks-ui';
import { useCheckoutStore, usePaymentIframe, useLineItems } from '@boldcommerce/checkout-react-components';
import { useAnalytics, useErrorLogging, useInventory, useMountedState } from '../../hooks';
import './CheckoutButton.css';
import { useTranslation } from 'react-i18next';

const CheckoutButton = ({ disabled, onClick, loading, className, errorMessage, innerRef }) => {
  const { t } = useTranslation();
  return (
  <>
    { errorMessage ? <Message type="alert">{ errorMessage }</Message> : null }
    <div ref={innerRef} className="CheckoutButton__Container">
      <Button onClick={onClick} disabled={disabled} loading={loading} className={className}>
        {t('complete_order')}
      </Button>
    </div>
  </>
)};

CheckoutButton.propTypes = {
  disabled: PropTypes.bool,
  onClick: PropTypes.func,
  loading: PropTypes.bool,
  className: PropTypes.string,
  errorMessage: PropTypes.string,
};

const CheckoutButtonContainer = ({ className }, ref) => {
  const { state } = useCheckoutStore();
  const [loading, setLoading] = useState();
  const { orderStatus } = state.orderInfo;
  const { processPaymentIframe } = usePaymentIframe();
  const checkInventory = useInventory();
  const { data: lineItems } = useLineItems();
  const history = useHistory();
  const trackEvent = useAnalytics();
  const logError = useErrorLogging();
  const isMounted = useMountedState();

  const orderErrorMessage = state.errors.order?.public_order_id;

  // don't disable checkout button if only error is order error
  const hasErrors = Object.keys(state.errors).some((errorKey) => errorKey != 'order' && errorKey != 'paymentIframe' && state.errors[errorKey] != null);
  const missingAddress = Object.values(state.applicationState.addresses).some(x => x === null || x.length === 0);
  const checkoutButtonDisabled = state.loadingStatus.isLoading || hasErrors ||  missingAddress;
  const processing = orderStatus === 'processing' || orderStatus === 'authorizing';
  const handleCheckout = async() => {
    setLoading(true);
    trackEvent('click_complete_order');
    try {
      const inventoryIssues = await checkInventory(lineItems);
      if (!inventoryIssues) {
        await processPaymentIframe();
        if(isMounted()) //need to check if mounted first to prevent memory leak
          setLoading(false);
      } else {
        setLoading(false);
        history.push('/inventory', inventoryIssues);
      }
    } catch(e) {  
      if(isMounted())
        setLoading(false);
      logError('checkout_button', e);
    }
  };
  const onClick = processing ? null : handleCheckout;

  return <CheckoutButton innerRef={ref} disabled={checkoutButtonDisabled} onClick={onClick} loading={processing || loading} className={className} errorMessage={orderErrorMessage} />;
};

const CheckoutButtonContainerForwardedRef = React.forwardRef(CheckoutButtonContainer);

export default CheckoutButtonContainerForwardedRef;
