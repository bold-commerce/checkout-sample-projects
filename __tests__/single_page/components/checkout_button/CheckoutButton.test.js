import React from "react";
import { render } from '@testing-library/react';
import { MemoryRouter } from "react-router-dom";
import { CheckoutButton } from '../../../../single_page/src/components/CheckoutButton';
import {
  exampleUseBillingAddress as MOCKbillingAddress,
  exampleUsePaymentIframe as MOCKPaymentIframe,
  exampleUseCheckoutStore as MOCKcheckoutStore,
  exampleUseLineItems as MOCKlineItems
} from "../../../utils/hookHelpers";
import '../../../../single_page/src/i18n/config';

jest.mock('@boldcommerce/checkout-react-components', () => ({
  ...jest.requireActual('@boldcommerce/checkout-react-components'),
  useBillingAddress: () => MOCKbillingAddress,
  usePaymentIframe: () => MOCKPaymentIframe,
  useCheckoutStore: () => MOCKcheckoutStore,
  useLineItems: () => MOCKlineItems,
}));


describe('CheckoutButton', () => {
  test('renders CheckoutButton component', () => {
    const { asFragment } = render(
      <CheckoutButton />,
      {wrapper: MemoryRouter}
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
