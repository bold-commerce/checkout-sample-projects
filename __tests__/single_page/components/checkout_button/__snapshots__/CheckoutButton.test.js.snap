// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckoutButton renders CheckoutButton component 1`] = `
<DocumentFragment>
  <button
    aria-busy="false"
    aria-disabled="false"
    class="stx-button Checkout__ConfirmButton"
    type="button"
  >
    <span
      class="stx-button__text"
    >
      Confirm Order
    </span>
    <span
      class="stx-loading-spinner stx-button__spinner"
    />
  </button>
</DocumentFragment>
`;
